client-device-connection-redis/src/main/java/org/eclipse/hono/deviceconnection/redis/client/VertxRedisCache.java

```diff
@@ -0,0 +1,338 @@
+        return Future.succeededFuture();
+    }
+
+    /*
+     * Will create a redis client and set up a reconnect handler when there is an exception in the connection.
+     */
+    private Future<RedisConnection> createRedisClient() {
+        final Promise<RedisConnection> promise = Promise.promise();
+
+        // make sure to invalidate old connection if present
+        if (redis != null) {
+            redis.close();
+        }
+
+        if (CONNECTING.compareAndSet(false, true)) {
+            redis = Redis.createClient(vertx, properties);
+            redis.connect()
+                    .onSuccess(conn -> {
+                        client = conn;
+
+                        // make sure the client is reconnected on error
+                        // eg, the underlying TCP connection is closed but the client side doesn't know it yet
+                        //     the client tries to use the staled connection to talk to server. An exceptions will be raised
+                        conn.exceptionHandler(e -> attemptReconnect(0));
+
+                        // make sure the client is reconnected on connection close
+                        // eg, the underlying TCP connection is closed with normal 4-Way-Handshake
+                        //     this handler will be notified instantly
+                        conn.endHandler(placeHolder -> attemptReconnect(0));
+
+                        // allow further processing
+                        promise.complete(conn);
+                        CONNECTING.set(false);
+                    }).onFailure(t -> {
+                        promise.fail(t);
+                        CONNECTING.set(false);
+                    });
+        } else {
+            promise.complete();
+        }
+
+        return promise.future();
+    }
+
+    private void attemptReconnect(final int retry) {
+        if (retry > MAX_RECONNECT_RETRIES) {
+            // we should stop now, as there's nothing we can do.
+            CONNECTING.set(false);
+        } else {
+            // retry with backoff up to 10240 ms
+            final long backoff = (long) (Math.pow(2, Math.min(retry, 10)) * 10);
+
+            vertx.setTimer(backoff, timer -> createRedisClient().onFailure(t -> attemptReconnect(retry + 1)));
+        }
+    }
+
+    @Override
+    public Future<JsonObject> checkForCacheAvailability() {
+        LOG.info("VREDIS: checking for cache availability");
+
+        Objects.requireNonNull(client);
+
+        return redis.connect().flatMap(connection -> {
+            final RedisAPI api = RedisAPI.api(connection);
+            return api.ping(List.of())
+                    .eventually(ignored -> connection.close())
+                    .mapEmpty();
+        });
+    }
+
+    @Override
+    public Future<Void> put(final String key, final String value) {
+        LOG.info("VREDIS: put {}={}", key, value);
+        Objects.requireNonNull(client);
+
+        return redis.connect().flatMap(connection -> {
```

Comment:
I assume that *connect()* immediately returns the connection if the client already is connected?

Reply:
I'm removing the call to connect and using RedisAPI directly. Still not sure whether it handles reconnects automatically but will dive deeper into that later.